File: src/kvald/synthetic_data.cpp
#include "kvald/synthetic_data.h"
#include <cmath>
#include <numeric>
#include <algorithm>

namespace kvald
{

Eigen::MatrixXf create_glare(
    int H,
    int W,
    const std::pair<int, int>& center,
    float sigma
)
{
    Eigen::MatrixXf g(H, W);
    int cy = center.first;
    int cx = center.second;

    for (int y = 0; y < H; ++y)
    {
        for (int x = 0; x < W; ++x)
        {
            g(y, x) = std::exp(-((std::pow(y - cy, 2) + std::pow(x - cx, 2)) / (2 * std::pow(sigma, 2))));
        }
    }
    return g / g.maxCoeff();
}

std::pair<std::vector<Eigen::MatrixXf>, std::vector<Eigen::MatrixXf>> static_glare(
    const std::vector<Eigen::MatrixXf>& background,
    std::mt19937& rng,
    float intensity,
    int sigma
)
{
    int T = background.size();
    if (T == 0) return {{}, {}};
    int H = background[0].rows();
    int W = background[0].cols();

    std::uniform_int_distribution<> dist_y(sigma, H - sigma - 1);
    std::uniform_int_distribution<> dist_x(sigma, W - sigma - 1);

    int cy = dist_y(rng);
    int cx = dist_x(rng);

    Eigen::MatrixXf spot = create_glare(H, W, {cy, cx}, sigma);

    std::vector<Eigen::MatrixXf> mask(T);
    std::vector<Eigen::MatrixXf> video(T);

    for (int t = 0; t < T; ++t)
    {
        mask[t] = spot.cast<float>();
        video[t] = (background[t].array() + intensity * mask[t].array()).cwiseMin(1.0f).cwiseMax(0.0f).matrix();
    }
    return {video, mask};
}

std::pair<std::vector<Eigen::MatrixXf>, std::vector<Eigen::MatrixXf>> pulsing_glare(
    const std::vector<Eigen::MatrixXf>& background,
    std::mt19937& rng,
    float intensity,
    int sigma
)
{
    int T = background.size();
    if (T == 0) return {{}, {}};
    int H = background[0].rows();
    int W = background[0].cols();

    std::uniform_int_distribution<> dist_y(sigma, H - sigma - 1);
    std::uniform_int_distribution<> dist_x(sigma, W - sigma - 1);
    std::uniform_int_distribution<> dist_period(20, 60);

    int cy = dist_y(rng);
    int cx = dist_x(rng);
    Eigen::MatrixXf spot = create_glare(H, W, {cy, cx}, sigma);
    int period = dist_period(rng);

    std::vector<Eigen::MatrixXf> mask(T);
    std::vector<Eigen::MatrixXf> video(T);

    for (int t = 0; t < T; ++t)
    {
        float alpha = 0.5f * (1.0f + std::sin(2 * M_PI * t / period));
        mask[t] = spot * alpha;
        video[t] = (background[t].array() + intensity * mask[t].array()).cwiseMin(1.0f).cwiseMax(0.0f).matrix();
    }
    return {video, mask};
}

std::pair<std::vector<Eigen::MatrixXf>, std::vector<Eigen::MatrixXf>> traveling_glare(
    const std::vector<Eigen::MatrixXf>& background,
    std::mt19937& rng,
    float intensity,
    int sigma
)
{
    int T = background.size();
    if (T == 0) return {{}, {}};
    int H = background[0].rows();
    int W = background[0].cols();

    std::uniform_int_distribution<> dist_y(sigma, H - sigma - 1);
    std::uniform_int_distribution<> dist_x(sigma, W - sigma - 1);

    std::pair<int, int> start = {dist_y(rng), dist_x(rng)};
    std::pair<int, int> end = {dist_y(rng), dist_x(rng)};

    std::vector<Eigen::MatrixXf> mask(T);
    std::vector<Eigen::MatrixXf> video(T);

    for (int t = 0; t < T; ++t)
    {
        float frac = static_cast<float>(t) / (T - 1.0f);
        int cy = static_cast<int>(start.first + frac * (end.first - start.first));
        int cx = static_cast<int>(start.second + frac * (end.second - start.second));
        Eigen::MatrixXf spot = create_glare(H, W, {cy, cx}, sigma);
        mask[t] = spot;
        video[t] = (background[t].array() + intensity * spot.array()).cwiseMin(1.0f).cwiseMax(0.0f).matrix();
    }
    return {video, mask};
}

std::pair<std::map<std::string, std::vector<Eigen::MatrixXf>>, std::map<std::string, std::vector<Eigen::MatrixXf>>> generate_sample_videos(
    int seed,
    int T,
    int H,
    int W,
    float mean,
    float std
)
{
    std::mt19937 rng(seed);
    std::normal_distribution<float> normal_dist(mean, std);

    std::vector<Eigen::MatrixXf> background(T);
    for (int t = 0; t < T; ++t)
    {
        background[t].resize(H, W);
        for (int r = 0; r < H; ++r)
        {
            for (int c = 0; c < W; ++c)
            {
                background[t](r, c) = std::min(1.0f, std::max(0.0f, normal_dist(rng)));
            }
        }
    }

    std::map<std::string, std::function<std::pair<std::vector<Eigen::MatrixXf>, std::vector<Eigen::MatrixXf>>(
        const std::vector<Eigen::MatrixXf>&, std::mt19937&, float, int)>> glare_funcs;
    glare_funcs["static"] = static_glare;
    glare_funcs["pulsing"] = pulsing_glare;
    glare_funcs["traveling"] = traveling_glare;

    std::string style = "pulsing"; // For consistent testing
    std::uniform_real_distribution<float> uniform_dist(0.5f, 1.0f);
    float intensity = uniform_dist(rng);
    std::uniform_int_distribution<> sigma_dist(1, 5);
    int sigma = sigma_dist(rng);

    auto result = glare_funcs[style](background, rng, intensity, sigma);
    
    std::map<std::string, std::vector<Eigen::MatrixXf>> video_dict;
    std::map<std::string, std::vector<Eigen::MatrixXf>> mask_dict;

    video_dict[style] = result.first;
    mask_dict[style] = result.second;

    return {video_dict, mask_dict};
}

} // namespace kvald

File: prototypes/synthetic_data.py
'''
A few key notes:
    I can use this as initial training data to start the model when I write it in C++, but I'm going to need some real data to actually train the model.
    I'll compile a text document full of driving footage, then I'll use a Python script to get the videos and plug them into the model.
    Initially I'll be using supervised learning, but this secondary step will be unsupervised so I'll need to figure out proper heuristics for the model.
'''

import numpy as np
from typing import Dict, Tuple
from torch.utils.data import Dataset
import torch


class GlareDataset(Dataset):
    def __init__(self, n_videos, transform=None):
        self.n_videos = n_videos
        self.transform = transform

    def __len__(self): return self.n_videos * 150 # Assuming 150 frames per video
    def __getitem__(self, idx):
        video_idx = idx // 150  # Assuming 150 frames per video
        frame_idx = idx % 150

        video_dict, mask_dict = generate_sample_videos(seed=video_idx)
        style = next(iter(video_dict))
        video = video_dict[style]  # (T,H,W)
        mask  = mask_dict[style]

        frame = video[frame_idx]
        mask  = mask[frame_idx]

        if self.transform:
            frame = self.transform(frame)
            mask  = self.transform(mask)
            
        return torch.from_numpy(frame)[None,...].float(), torch.from_numpy(mask)[None,...].float()  # shape (1,H,W), (1,H,W)

def create_glare(
        H: int, 
        W: int, 
        center: Tuple[int, int], 
        sigma: float
) -> np.ndarray:
    """2D Gaussian spot of shape (H, W), normalized to [0,1]."""
    y = np.arange(H)[:, None]
    x = np.arange(W)[None, :]
    cy, cx = center
    g = np.exp(-(((y - cy) ** 2 + (x - cx) ** 2) / (2 * sigma ** 2)))
    return g / g.max()

# Apply a static Gaussian glare at a random position.
def static_glare(
    background: np.ndarray,
    rng: np.random.RandomState,
    intensity: float,
    sigma: int
) -> Tuple[np.ndarray, np.ndarray]:
    
    T, H, W = background.shape
    cy = rng.randint(sigma, H - sigma)
    cx = rng.randint(sigma, W - sigma)
    spot = create_glare(H, W, (cy, cx), sigma)
    mask = np.tile(spot, (T, 1, 1)).astype(np.float32)
    video = np.clip(background + intensity * mask, 0, 1)
    return video, mask

# Apply a pulsing Gaussian glare at a random position with random period.
def pulsing_glare(
    background: np.ndarray,
    rng: np.random.RandomState,
    intensity: float,
    sigma: int
) -> Tuple[np.ndarray, np.ndarray]:
    
    T, H, W = background.shape
    cy = rng.randint(sigma, H - sigma)
    cx = rng.randint(sigma, W - sigma)
    spot = create_glare(H, W, (cy, cx), sigma)
    period = rng.randint(20, 60)
    mask = np.zeros_like(background)
    for t in range(T):
        alpha = 0.5 * (1 + np.sin(2 * np.pi * t / period))
        mask[t] = spot * alpha
    video = np.clip(background + intensity * mask, 0, 1)
    return video, mask.astype(np.float32)

# Apply a traveling Gaussian glare moving linearly between random start/end points.
def traveling_glare(
    background: np.ndarray,
    rng: np.random.RandomState,
    intensity: float,
    sigma: int
) -> Tuple[np.ndarray, np.ndarray]:

    T, H, W = background.shape
    start = (rng.randint(sigma, H - sigma), rng.randint(sigma, W - sigma))
    end = (rng.randint(sigma, H - sigma), rng.randint(sigma, W - sigma))
    mask = np.zeros_like(background)
    video = np.zeros_like(background)
    for t in range(T):
        frac = t / float(T - 1)
        cy = int(start[0] + frac * (end[0] - start[0]))
        cx = int(start[1] + frac * (end[1] - start[1]))
        spot = create_glare(H, W, (cy, cx), sigma)
        mask[t] = spot
        video[t] = np.clip(background[t] + intensity * spot, 0, 1)
    return video, mask.astype(np.float32)

# Generate one random glare style over a Gaussian background.
def generate_sample_videos(
    seed: int,
    T: int = 150,
    H: int = 128,
    W: int = 128,
    mean: float = 0.5,
    std: float = 0.1
) -> Tuple[Dict[str, np.ndarray], Dict[str, np.ndarray]]:
    """
    Returns:
      video: single-key dict {style: (T,H,W) float32 in [0,1]}
      mask:  single-key dict {style: (T,H,W) float32 ground-truth}
    """
    rng = np.random.RandomState(seed)
    background = rng.normal(loc=mean, scale=std, size=(T, H, W)).astype(np.float32)
    background = np.clip(background, 0.0, 1.0)

    glare_funcs = { # can add more if necessary
        'static': static_glare,
        'pulsing': pulsing_glare,
        'traveling': traveling_glare,
    }
    style = "pulsing" # For consistent testing
    intensity = float(rng.uniform(0.5, 1.0))
    sigma = int(rng.uniform(1, 5))

    video, mask = glare_funcs[style](background, rng, intensity, sigma)
    return {style: video}, {style: mask}

if __name__ == '__main__':
    import json
    import os

    # Create the data directory if it doesn't exist
    data_dir = os.path.join(os.path.dirname(__file__), '..', 'data')
    if not os.path.exists(data_dir):
        os.makedirs(data_dir)

    video_dict, mask_dict = generate_sample_videos(seed=0, T=10, H=16, W=16)
    style = next(iter(video_dict))
    video = video_dict[style]
    mask = mask_dict[style]

    data_to_serialize = {
        'style': style,
        'video': video.tolist(),
        'mask': mask.tolist()
    }

    with open(os.path.join(data_dir, 'synthetic_data.json'), 'w') as f:
        json.dump(data_to_serialize, f, indent=4)